Source/WebCore/rendering/RenderLayer.cpp

```diff
@@ -5829,26 +5829,54 @@ static bool rendererHasHDRContent(const RenderElement& renderer)
             if (cachedImage->hasHDRContent())
                 return true;
         }
-        return false;
-    }
-
-    if (CheckedPtr imageRenderer = dynamicDowncast<LegacyRenderSVGImage>(renderer)) {
+    } else if (CheckedPtr imageRenderer = dynamicDowncast<LegacyRenderSVGImage>(renderer)) {
         if (auto* cachedImage = imageRenderer->imageResource().cachedImage()) {
             if (cachedImage->hasHDRContent())
                 return true;
         }
-        return false;
-    }
-
 #if ENABLE(PIXEL_FORMAT_RGBA16F)
-    if (CheckedPtr canavsRenderer = dynamicDowncast<RenderHTMLCanvas>(renderer)) {
+    } else if (CheckedPtr canavsRenderer = dynamicDowncast<RenderHTMLCanvas>(renderer)) {
         if (auto* renderingContext = canavsRenderer->canvasElement().renderingContext()) {
             if (renderingContext->isHDR())
                 return true;
         }
+#endif
+    }
+
+    auto styleHasHDRContent = [](const auto* style) {
+        if (!style)
+            return false;
+
+        if (style->hasBackgroundImage()) {
+            if (style->backgroundLayers().hasHDRContent())
+                return true;
+        }
+
+        if (style->hasBorderImage()) {
+            auto image = style->borderImage().image();
+            if (auto* cachedImage = image ? image->cachedImage() : nullptr) {
+                if (cachedImage->hasHDRContent())
+                    return true;
+            }
+        }
+
+        if (auto image = style->listStyleImage()) {
+            if (auto* cachedImage = image->cachedImage()) {
+                if (cachedImage->hasHDRContent())
+                    return true;
+            }
+        }
+
         return false;
+    };
+
+    if (styleHasHDRContent(&renderer.style()))
+        return true;
+
+    for (auto pseudoId = PseudoId::FirstPublicPseudoId; pseudoId < PseudoId::FirstInternalPseudoId; pseudoId = static_cast<PseudoId>(static_cast<unsigned>(pseudoId) + 1)) {
+        if (styleHasHDRContent(renderer.getCachedPseudoStyle({ pseudoId })))
+            return true;
     }
```

Comment:
Why are we doing this? Shouldn't pseudo-elements be like normal element children?

Also the iteration here won't work properly when targeting pseudo-elements with arguments like highlights or view transitions.